src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs

```diff
@@ -27,75 +27,103 @@ internal class CookieAuthenticationHandler : AuthenticationHandler<CookieAuthent
+                        if (ticket == null)
+                        {
+                            Logger.LogWarning(@"Identity missing in session store");
+                            return;
+                        }
                     }
-                    return null;
-                }
 
-                var allowRefresh = ticket.Properties.AllowRefresh ?? true;
-                if (issuedUtc != null && expiresUtc != null && Options.SlidingExpiration && allowRefresh)
-                {
-                    var timeElapsed = currentUtc.Subtract(issuedUtc.Value);
-                    var timeRemaining = expiresUtc.Value.Subtract(currentUtc);
+                    var currentUtc = Options.SystemClock.UtcNow;
+                    var issuedUtc = ticket.Properties.IssuedUtc;
+                    var expiresUtc = ticket.Properties.ExpiresUtc;
 
-                    if (timeRemaining < timeElapsed)
+                    if (expiresUtc != null && expiresUtc.Value < currentUtc)
                     {
-                        _shouldRenew = true;
-                        _renewIssuedUtc = currentUtc;
-                        var timeSpan = expiresUtc.Value.Subtract(issuedUtc.Value);
-                        _renewExpiresUtc = currentUtc.Add(timeSpan);
+                        if (Options.SessionStore != null)
+                        {
+                            await Options.SessionStore.RemoveAsync(_sessionKey);
+                        }
+                        return;
                     }
+
+                    var allowRefresh = ticket.Properties.AllowRefresh ?? true;
+                    if (issuedUtc != null && expiresUtc != null && Options.SlidingExpiration && allowRefresh)
+                    {
+                        var timeElapsed = currentUtc.Subtract(issuedUtc.Value);
+                        var timeRemaining = expiresUtc.Value.Subtract(currentUtc);
+
+                        if (timeRemaining < timeElapsed)
+                        {
+                            _shouldRenew = true;
+                            _renewIssuedUtc = currentUtc;
+                            var timeSpan = expiresUtc.Value.Subtract(issuedUtc.Value);
+                            _renewExpiresUtc = currentUtc.Add(timeSpan);
+                        }
+                    }
+
+                    // Finally we have a valid ticket
+                    _cookieTicket = ticket;
+                }
+                finally
+                {
+                    _cookieRead = true;
                 }
+            }
+        }
 
-                var context = new CookieValidatePrincipalContext(Context, ticket, Options);
+        protected override async Task<AuthenticationTicket> HandleAuthenticateAsync()
+        {
+            AuthenticationTicket ticket = null;
+            try
+            {
+                await EnsureCookieTicket();
+                ticket = _cookieTicket;
```

Comment:
You mean the pattern in the base class?  Its for the exception handler which sometimes changes the ticket... for some of these methods...  Yes I noticed this code looks bad :)  Let me take a look again